lib/ansible/modules/file.py

```diff
@@ -552,35 +552,59 @@ def execute_touch(path, follow, timestamps):
     result = {'dest': path}
     mtime = get_timestamp_for_time(timestamps['modification_time'], timestamps['modification_time_format'])
     atime = get_timestamp_for_time(timestamps['access_time'], timestamps['access_time_format'])
+    diff = initial_diff(path, 'touch', prev_state)
+    file_args = module.load_file_common_arguments(module.params)
 
-    if not module.check_mode:
-        if prev_state == 'absent':
-            # Create an empty file if the filename did not already exist
+    # If the file did not already exist
+    if prev_state == 'absent':
+        # if we are in check mode and the file is absent
+        # we can set the changed status to True.
+        if module.check_mode:
+            result['changed'] = True
+            return result
+        else:
+            # Create an empty file
             try:
                 open(b_path, 'wb').close()
                 changed = True
             except (OSError, IOError) as e:
                 raise AnsibleModuleError(results={'msg': 'Error, could not touch target: %s'
-                                                         % to_native(e, nonstring='simplerepr'),
+                                                  % to_native(e, nonstring='simplerepr'),
                                                   'path': path})
 
-        # Update the attributes on the file
-        diff = initial_diff(path, 'touch', prev_state)
-        file_args = module.load_file_common_arguments(module.params)
-        try:
+    # Update the attributes on the file
+    try:
+        # If we are in check mode, check if the file attributes are changing. We call the same
+        # functions used by the set_fs_attributes_if_different, except for the user and group for which the call
+        # will fail if the user/group does not yet exist
+        if module.check_mode:
+            # get current file attributes
+            file = os.stat(path)
+            # check if user is different (name used since it could not yet exist)
+            changed = file_args['owner'] is not None and file_args['owner'] != getpwuid(file.st_uid).pw_name
```

Comment:
This shouldn't be unique to `state=touch`. In main() the module checks if the owner/group exist specifically to give a warning in check mode, so you could reuse that information instead of guessing that owner is a name rather than uuid. If the group/owner don't exist this function could exit early (or not be called), so then check mode/non check mode can both just use `set_fs_attributes_if_different`.

Reply:
I'm not sure considering a missing group/owner as changed makes sense - playbooks should use conditionals to only run task when necessary setup is complete, not use check mode for that.

Reply:
Hi @s-hertel , thank you for the feedback! I understand your point, just in that case if we early return from that function we will obtain the `changed` property to be `false` and not `true`, so the overall task won't be shown as changed. Do you think this is still fine? If so I will proceed immediately by working on this.

Reply:
`check: true` on a task should return the same `changed` status as `check: false`, that is the purpose of it. If you return early you'd need to make sure the returned data is correct.

The point was to not make a `state=touch` specific quirk (have you tested `state=file` with check mode and a group that does not exist?) and avoid reimplementing `set_fs_attributes_if_different` in the file module since the method works with check mode, except if the group/owner doesn't exist.

Stepping back from the implementation though, I don't think this is the right behavior for non-existent owner/groups in check mode. The task doesn't know if a previous task would have created it, so failing seems more cautious - better than the user having the expectation that the playbook will succeed and then running without check mode to discover an error. Rather than expecting check mode to work across tasks, tasks should use conditionals imo.

Reply:
No I perfectly agree with you @s-hertel and thank you for your opinion. Honestly I was in doubts between the two approaches when thinking about how this should behave, and obviously I chose the wrong one. I just pushed a revised version which in the end I feel simpler and cleaner as well, please let me know how I can furtherly improve it and thank you for your time and help!